### ISSUE TITLE
Each Menu adds its own document event listeners for detecting keyboard state

### ISSUE THREAD

**Issue by AlabasterAxe** [radix-ui/primitives#1239]
## Bug report

This is basically a reopening of https://github.com/radix-ui/primitives/issues/1139 since we're still seeing considerable slow down even after updating past the fix.

### Current Behavior

Radix Dropdown Menus and Context Menus each add their own document event listeners for keydown and pointerdown which can cause a slowdown in the handling of the corresponding events

### Expected behavior

Handling of keydown and pointerdown don't slowdown as a result of additional menu components on the page.

### Reproducible example

Reproduction is to add n menus to a react app and observe that there are n document event listeners for either keydown or pointerdown

My Profile demonstrating consistent slowness with handling mousedown events:
[Profile-20220312T073820.json.zip](https://github.com/radix-ui/primitives/files/8237628/Profile-20220312T073820.json.zip)

### Suggested solution

Provide a global keyboard state context which registers a single document event listener and all menus can share.

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) |   @radix-ui/react-context-menu, @radix-ui/react-dropdown-menu      |    0.1.5     |
| React            | n/a     |     16.13    |
| Browser          |    Chrome     |     99    |
| Assistive tech   |    n/a     |   n/a      |
| Node             | n/a     |         |
| npm/yarn         |    yarn     |     1.22    |
| Operating System |     MacOS    |    12.2.1     |

**Comment by AlabasterAxe** [comment]
@benoitgrelard @jjenzz Just pinging this. This is causing serious overhead in our app. It would be great if we could have a solution in Radix that didn't incur a performance penalty for each additional menu. Thanks!

**Comment by jjenzz** [comment]
@hipstersmoothie many of the components listed would only add events when the content parts are open so unlikely to be an issue unless you have 100s of dialogs open at the same time (for example).

the issue with `Menu` is specifically because this one is adding events in the `Root` part which will exist for all instances on the page (not just open content). the only one from your list that i can see is likely to suffer the same issues is `RadioGroup`.

**Comment by jjenzz** [comment]
@AlabasterAxe i'm not working at modulz anymore so i cannot help with a timeline for a solution on this i'm afraid.

**Comment by AlabasterAxe** [comment]
Nice thanks for the pointer! taking a look. I agree that example performs very well. Not sure why performance would be so different.

In the cases I've been looking at it was ~700.

**Comment by AlabasterAxe** [comment]
Your page does reproduce the same issue:
<img width="814" alt="Screen Shot 2022-03-16 at 7 01 14 PM" src="https://user-images.githubusercontent.com/1403638/158705154-ee58175f-0c59-4120-89ff-98531707120f.png">

The top-level task took 42 ms overall and 40ms of which is just handling the keydown event listeners. It's possible that if there's more going on on the page that these event handlers are slower to process? 🤷

**Comment by hipstersmoothie** [comment]
Yeah this is easily reproducible with the given example. Here I upped the menus to 1000 and each keydown takes 140ms

![CleanShot 2022-03-16 at 16 23 56](https://user-images.githubusercontent.com/1192452/158707470-18adc41b-eedf-4507-8e17-9b6ef8c3b869.png)

Even downing it to 700 (matching our usage) see `80ms` delays

![CleanShot 2022-03-16 at 16 25 02](https://user-images.githubusercontent.com/1192452/158707641-99475d41-feb0-490f-a97d-c1054368549a.png)

[This PR I opened](https://github.com/radix-ui/primitives/pull/1252) does fix the issue though. 🙏ing we can get a merge

**Comment by jjenzz** [comment]
> Here I upped the menus to 1000 and each keydown takes 140ms

while i agree it might be nice to remove these events (to prevent the overhead of these reports if anything), if you're experiencing a performance issue from them i would think you are rendering far too many menus at once.

personally i would avoid rendering 500 menus on screen let alone 1000, since it is unlikely to be possible for a human to interact with that many menus in a single viewport so react will be redundantly managing those trees. 

using something like IntersectionObserver/virtual scroll to render only the menus in the users viewport will help.

**Comment by jjenzz** [comment]
> using something like IntersectionObserver/virtual scroll to render only the menus in the user's viewport will help your performance all round.

@benoitgrelard @andy-hook  that suggestion actually just made me realise that the `Menu` could probably use `IntersectionObserver` to verify it's in view before binding the events.

**Comment by AlabasterAxe** [comment]
@jjenzz 100% agree, there's no reason we need to render many of those components since most of them are off-screen. I'm planning on looking into this soon.

At the same time, I think it's reasonable for these components to defensive about potential misuse, especially since in this case it seems like there are reasonable ways to avoid registering one listener per menu.

**Comment by benoitgrelard** [comment]
Hi @AlabasterAxe,

> At the same time, I think it's reasonable for these components to defensive about potential misuse, especially since in this case it seems like there are reasonable ways to avoid registering one listener per menu.

Whilst this might be nice, it is clear that a lot more can be done user-side/app-side to ensure better performance.
Additionally, we have already reduced the numbers of listeners for each menu instance from 3 to 1 (#1140).

We have to prioritise where we focus our time so I will close this issue for now as it is unlikely we will do more in this case.

✌️
